src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -2018,6 +2018,72 @@ BOOST_AUTO_TEST_CASE(GradientAtrousConvolutionLayerTest)
+{
+  // Parameter order for the constructor: inSize, outSize, kW, kH, dW, dH, padW,
+  // padH, inputWidth, inputHeight, dilationW, dilationH, paddingType ("none").
+  AtrousConvolution<> layer1(1, 2, 3, 4, 5, 6, std::make_tuple(7, 8),
+      std::make_tuple(9, 10), 11, 12, 13, 14);
+  AtrousConvolution<> layer2(2, 3, 4, 5, 6, 7, std::make_tuple(8, 9),
+      std::make_tuple(10, 11), 12, 13, 14, 15);
+
+  // Make sure we can get the parameters successfully.
+  BOOST_REQUIRE_EQUAL(layer1.InputWidth(), 11);
+  BOOST_REQUIRE_EQUAL(layer1.InputHeight(), 12);
+  BOOST_REQUIRE_EQUAL(std::get<0>(layer1.KernelSize()), 3);
+  BOOST_REQUIRE_EQUAL(std::get<1>(layer1.KernelSize()), 4);
+  BOOST_REQUIRE_EQUAL(std::get<0>(layer1.Strides()), 5);
+  BOOST_REQUIRE_EQUAL(std::get<1>(layer1.Strides()), 6);
+  BOOST_REQUIRE_EQUAL(std::get<0>(layer1.PadHSize()), 9);
+  BOOST_REQUIRE_EQUAL(std::get<1>(layer1.PadHSize()), 10);
+  BOOST_REQUIRE_EQUAL(std::get<0>(layer1.PadWSize()), 7);
+  BOOST_REQUIRE_EQUAL(std::get<1>(layer1.PadWSize()), 8);
+  BOOST_REQUIRE_EQUAL(std::get<0>(layer1.DilationRate()), 13);
+  BOOST_REQUIRE_EQUAL(std::get<1>(layer1.DilationRate()), 14);
+
+  // Now modify the parameters to match the second layer.
+  layer1.InputWidth() = 12;
+  layer1.InputHeight() = 13;
+  std::get<0>(layer1.KernelSize()) = 4;
+  std::get<1>(layer1.KernelSize()) = 5;
+  std::get<0>(layer1.Strides()) = 6;
+  std::get<1>(layer1.Strides()) = 7;
+  std::get<0>(layer1.PadHSize()) = 10;
+  std::get<1>(layer1.PadHSize()) = 11;
+  std::get<0>(layer1.PadWSize()) = 8;
+  std::get<1>(layer1.PadWSize()) = 9;
+  std::get<0>(layer1.DilationRate()) = 12;
+  std::get<1>(layer1.DilationRate()) = 13;
+
+  // Now ensure all results are the same.
+  BOOST_REQUIRE_EQUAL(layer1.InputWidth(), layer2.InputWidth());
+  BOOST_REQUIRE_EQUAL(layer1.InputHeight(), layer2.InputHeight());
+  BOOST_REQUIRE_EQUAL(std::get<0>(layer1.KernelSize()),
+                      std::get<0>(layer2.KernelSize()));
+  BOOST_REQUIRE_EQUAL(std::get<1>(layer1.KernelSize()),
+                      std::get<1>(layer2.KernelSize()));
+  BOOST_REQUIRE_EQUAL(std::get<0>(layer1.Strides()),
+                      std::get<0>(layer2.Strides()));
+  BOOST_REQUIRE_EQUAL(std::get<1>(layer1.Strides()),
+                      std::get<1>(layer2.Strides()));
+  BOOST_REQUIRE_EQUAL(std::get<0>(layer1.PadHSize()),
+                      std::get<0>(layer2.PadHSize()));
+  BOOST_REQUIRE_EQUAL(std::get<1>(layer1.PadHSize()),
+                      std::get<1>(layer2.PadHSize()));
+  BOOST_REQUIRE_EQUAL(std::get<0>(layer1.PadWSize()),
+                      std::get<0>(layer2.PadWSize()));
+  BOOST_REQUIRE_EQUAL(std::get<1>(layer1.PadWSize()),
+                      std::get<1>(layer2.PadWSize()));
+  BOOST_REQUIRE_EQUAL(std::get<0>(layer1.DilationRate()),
+                      std::get<0>(layer1.DilationRate()));
+  BOOST_REQUIRE_EQUAL(std::get<1>(layer2.DilationRate()),
+                      std::get<1>(layer2.DilationRate()));
```

Comment:
I am of the view that the we should also check that the `Forward` and `Backward` functions are also working as expected. Let me know what you think.

Reply:
If all the parameters are the same, would we expect the outputs of those functions to be different?  I can add a test if you like, I'm just not sure I see the need.  Maybe there is something I am overlooking though.

Reply:
Ah yes as we will be dealing with the `reset` issue afterward. I think it would be fine if we leave this as it is :+1:

Reply:
Ahh, I see, really it is the `Reset()` here that is the issue.  Anyway, yeah, I agree, we can leave it as is for now and deal with it later.